### PR TITLE
Log send mobile

### PR DESCRIPTION
**Pull request by MarcoPolo**
👓  @keybase/react-hackers 

This implements the Android native modules and UI components (ios+android) to send a log send

It's not pretty, but it works.

This has minimal go changes so the actual log send doesn't work. Those changes will be in a separate PR.

### Previews

screen shot 2016-05-11 at 5 10 22 pm
screen shot 2016-05-11 at 5 10 34 pm
screen shot 2016-05-11 at 5 10 32 pm

### CODE REVIEW

File: react-native/android/app/build.gradle [hunk 1]
```diff
@@ -156,6 +156,12 @@ dependencies {
     compile "com.facebook.react:react-native:+"  // From node_modules
     compile project(':ReactNativeBarcodescanner')
     compile project(':RNPermissionsModule')
+    compile 'org.slf4j:slf4j-api:1.7.21'
+    compile 'com.github.tony19:logback-android-core:1.1.1-5'
+    compile('com.github.tony19:logback-android-classic:1.1.1-5') {
+        // workaround issue #73
```
Comment (chromakode) [inline react-native/android/app/build.gradle]: This comment is unclear to me -- issue 73 of what?

Comment (MarcoPolo) [inline react-native/android/app/build.gradle]: I'll update, it means: https://github.com/tony19/logback-android/issues/73

File: react-native/android/app/src/main/java/io/keybase/ossifrage/modules/FileLogger.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    public FileLogger(final ReactApplicationContext reactContext, String logFilePath) {
+        super(reactContext);
+        configureLogbackDirectly(logFilePath);
+        log = LoggerFactory.getLogger(MainActivity.class);
+    }
+
+    synchronized public static void configureLogbackDirectly(String filePath) {
+        if (FileLogger.configured) {
+            return;
```
Comment (chromakode) [inline react-native/android/app/src/main/java/io/keybase/ossifrage/modules/FileLogger.java]: In what case might this be called a second time?

Comment (MarcoPolo) [inline react-native/android/app/src/main/java/io/keybase/ossifrage/modules/FileLogger.java]: Whenever you reload js you recreate your native modules.

This is also a static method, and may be helpful to be used outside of this module. (Say configuring from MainActivity, which I'm currently not doing.)

File: shared/more/log-send/index.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    return <Render logSendId={this.state.logSendId} onLogSend={onLogSend} />
+  }
+
+  static parseRoute () {
+    return {componentAtTop: {title: 'Log Send'}}
+  }
+}
+
+export default connect()(LogSend)
```
Comment (chromakode) [inline shared/more/log-send/index.js]: Does this need to be `connect`ed?

Comment (chromakode) [inline shared/more/log-send/index.js]: Ah, for `parseRoute`. Nvm.

Comment (chromakode) [inline shared/more/log-send/index.js]: Er, second guessing myself again. :) That's not related, is it?

Comment (MarcoPolo) [inline shared/more/log-send/index.js]: I don't think it needs to be, I probably copied some code somewhere that had that.

File: shared/more/log-send/render.desktop.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// @flow
+import React, {Component} from 'react'
+import {Box, Text} from '../../common-adapters'
+
+export default class ChatRender extends Component {
```
Comment (chromakode) [inline shared/more/log-send/render.desktop.js]: s/ChatRender/Render?

File: shared/more/log-send/render.native.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  render () {
+    const onSubmitIssue = () => {
+      Linking.openURL(`https://github.com/keybase/client/issues/new?body=[write%20something%20useful%20and%20descriptive%20here]%0A%0Amy%20log%20id:%20${this.props.logSendId}`)
+    }
+
+    if (!this.props.logSendId) {
+      return (
+        <Box style={stylesContainer}>
+          <Text type='Body'>Send a log Send?</Text>
+          <Text type='Body' style={stylesInfoText}>This command will send recent keybase log entries to keybase.io for debugging purposes only.  These logs don’t include your private keys or encrypted data, but they will include filenames and other metadata keybase normally can’t read, for debugging purposes.</Text>
```
Comment (chromakode) [inline shared/more/log-send/render.native.js]: Capitalize "keybase"?

File: shared/more/log-send/render.native.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+  render () {
+    const onSubmitIssue = () => {
+      Linking.openURL(`https://github.com/keybase/client/issues/new?body=[write%20something%20useful%20and%20descriptive%20here]%0A%0Amy%20log%20id:%20${this.props.logSendId}`)
+    }
+
+    if (!this.props.logSendId) {
+      return (
+        <Box style={stylesContainer}>
+          <Text type='Body'>Send a log Send?</Text>
```
Comment (chromakode) [inline shared/more/log-send/render.native.js]: This sentence is a bit confusing :) "Send debug logs"?

File: shared/more/log-send/render.native.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+      return (
+        <Box style={stylesContainer}>
+          <Text type='Body'>Your log id is:</Text>
+          <Text type='Terminal'
+            onClick={() => this._copyToClipboard()}>
+            {this.props.logSendId} (tap to copy)
+          </Text>
+          {this.state.copiedToClipboard && <Text type='Body'>Copied to clipboard!</Text>}
+
+          <Text type='Body'>Send us the log id along with your problem in this github issue:</Text>
```
Comment (chromakode) [inline shared/more/log-send/render.native.js]: s/your problem/a description of what's going on/?
s/github/GitHub/?

File: shared/more/log-send/render.native.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+          <Button type='Primary' label='File a github issue:' onClick={onSubmitIssue} />
```
Comment (chromakode) [inline shared/more/log-send/render.native.js]: s/github/GitHub

File: shared/more/log-send/render.native.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    const onSubmitIssue = () => {
+      Linking.openURL(`https://github.com/keybase/client/issues/new?body=[write%20something%20useful%20and%20descriptive%20here]%0A%0Amy%20log%20id:%20${this.props.logSendId}`)
+    }
+
+    if (!this.props.logSendId) {
+      return (
+        <Box style={stylesContainer}>
+          <Text type='Body'>Send a log Send?</Text>
+          <Text type='Body' style={stylesInfoText}>This command will send recent keybase log entries to keybase.io for debugging purposes only.  These logs don’t include your private keys or encrypted data, but they will include filenames and other metadata keybase normally can’t read, for debugging purposes.</Text>
+          <Button type='Primary' label='Send a Log!' onClick={this.props.onLogSend} />
```
Comment (chromakode) [inline shared/more/log-send/render.native.js]: Uncapitalize log here?

File: shared/native/log-send.desktop.js [hunk 10]
```diff
@@ -0,0 +1 @@
+export default function () { return Promise.resolve('none') }
```
Comment (chromakode) [inline shared/native/log-send.desktop.js]: Why do these stub functions return a promise?

**Comment by chromakode** [comment]
:+1: Sorry if this is too many nitpicks. This was really interesting to read!

File: shared/native/log-send.desktop.js (hunk 10 above)
Comment (MarcoPolo) [inline shared/native/log-send.desktop.js]: That's the flow type we have to fulfill

File: shared/more/log-send/render.native.js (hunk 6 above)
Comment (MarcoPolo) [inline shared/more/log-send/render.native.js]: So close to a palindrome!

File: shared/more/log-send/render.desktop.js (hunk 4 above)
Comment (MarcoPolo) [inline shared/more/log-send/render.desktop.js]: Damnit! I fixed that in a different file :P

File: shared/native/log-send.desktop.js (hunk 10 above)
Comment (chromakode) [inline shared/native/log-send.desktop.js]: Gotcha, thanks.

File: react-native/android/app/src/main/java/io/keybase/ossifrage/modules/FileLogger.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
+package io.keybase.ossifrage.modules;
```
Comment (chrisnojima) [inline react-native/android/app/src/main/java/io/keybase/ossifrage/modules/FileLogger.java]: can we talk about why this is necessary? go should be able to write logs locally w/o this

Comment (MarcoPolo) [inline react-native/android/app/src/main/java/io/keybase/ossifrage/modules/FileLogger.java]: This is akin to this: https://github.com/keybase/client/blob/master/desktop/app/console-helper.js

but for android.

It's not for the go side, it's to forward logs from JS to android and have android log it to a file

Comment (MarcoPolo) [inline react-native/android/app/src/main/java/io/keybase/ossifrage/modules/FileLogger.java]: specifically the part that write to the fs

Comment (chrisnojima) [inline react-native/android/app/src/main/java/io/keybase/ossifrage/modules/FileLogger.java]: 👍

**Comment by MarcoPolo** [comment]
@chrisnojima anything else?

**Comment by chrisnojima** [comment]
👍 but we should make a ticket for the ios version

**Pull request merged**